numba/tests/test_typeinfer.py

```diff
@@ -759,5 +763,75 @@ def test_fold_arguments_list_inputs(self):
                 self.check_fold_arguments_list_inputs(**case)
 
 
+@register_pass(mutates_CFG=False, analysis_only=True)
+class DummyCR(FunctionPass):
+    """Dummy pass to add "cr" to compiler state to avoid errors in TyperCompiler since
+    it doesn't have lowering.
+    """
+
+    _name = "dummy_cr"
+
+    def __init__(self):
+        FunctionPass.__init__(self)
+
+    def run_pass(self, state):
+        state.cr = 1  # arbitrary non-None value
+        return True
+
+
+class TyperCompiler(numba.core.compiler.CompilerBase):
+    """A compiler pipeline that skips passes after typing (provides partial typing info
+    but not lowering).
+    """
+
+    def define_pipelines(self):
+        pm = numba.core.compiler_machinery.PassManager("custom_pipeline")
+        pm.add_pass(TranslateByteCode, "analyzing bytecode")
+        pm.add_pass(IRProcessing, "processing IR")
+        pm.add_pass(PartialTypeInference, "do partial typing")
+        pm.add_pass_after(DummyCR, PartialTypeInference)
+        pm.finalize()
+        return [pm]
+
+
+def get_func_typing_errs(func, arg_types):
+    """
+    Get typing errors for function 'func'. It creates a pipeline that runs untyped
+    passes as well as type inference.
+    """
+    typingctx = numba.core.registry.cpu_target.typing_context
+    targetctx = numba.core.registry.cpu_target.target_context
+    library = None
+    return_type = None
+    _locals = {}
+    flags = numba.core.compiler.Flags()
+
+    pipeline = TyperCompiler(
+        typingctx, targetctx, library, arg_types, return_type, flags, _locals
+    )
+    pipeline.compile_extra(func)
+    return pipeline.state.typing_errors
+
+
+class TestPartialTypingErrors(unittest.TestCase):
+    """
+    Make sure partial typing stores type errors in compiler state properly
+    """
+
+    def test_partial_typing_error(self):
+        # example with type unification error
+        def impl(flag):
+            if flag:
+                a = 1
+            else:
+                a = np.ones(3)
+            return a
+
+        typing_errs = get_func_typing_errs(impl, (types.bool_,))
+        self.assert_(isinstance(typing_errs, list) and len(typing_errs) == 1)
+        self.assert_(isinstance(typing_errs[0], errors.TypingError)
```

Comment:
`assert_` is a deprecated alias. Use `assertTrue` instead.

Reply:
Done.